File: Data Structures & Algorithms/combination-target-sum/submission-2.cpp
class Solution {
public:
    void backtrack(vector<int>& nums, int target) {
    int sum = 0;
    for (auto& t : temp) sum += t;
    if (sum >= target) {
        if (sum == target && find(counters.begin(), counters.end(), temp_counter) == counters.end()) {
             result.push_back(temp);
             counters.push_back(temp_counter);
        }
        return;
    }

    for (int i = 0; i < nums.size(); ++i) {
        temp.push_back(nums[i]);
        temp_counter[nums[i]]++;
        backtrack(nums, target);
        temp_counter[nums[i]]--;
        temp.pop_back();
    }
}

    vector<vector<int>> combinationSum(vector<int>& nums, int target) {
        for (auto& num : nums) {
            temp_counter[num] = 0;
        }
        backtrack(nums, target);
        return result;
    }

    vector<unordered_map<int, int>> counters;
    unordered_map<int, int> temp_counter;
    vector<int> temp;
    vector<vector<int>> result;
};

File: Data Structures & Algorithms/find-target-in-rotated-sorted-array/submission-2.py
class Solution:
    def search(self, nums: List[int], target: int) -> int:
        l, r = 0, len(nums) - 1

        while l < r:
            middle_idx = l + int((r - l) / 2)
            if nums[middle_idx] == target: return middle_idx
            elif nums[l] == target: return l
            elif nums[r] == target: return r
            if nums[l] <= nums[middle_idx]:
                if target < nums[middle_idx] and target > nums[l]: r = middle_idx - 1
                elif target > nums[middle_idx] and target > nums[r]: l = middle_idx + 1
                elif target > nums[middle_idx] or target < nums[l]: l = middle_idx + 1
                else: return -1
            elif nums[middle_idx] < nums[r]:
                if target > nums[middle_idx] and target < nums[r]: l = middle_idx + 1
                elif target > nums[middle_idx] and target > nums[r]: r = middle_idx - 1
                elif target > nums[r] or target < nums[middle_idx]: r = middle_idx - 1
                else: return -1

        return l if nums[l] == target else -1